Fix multi-arch cross-compiling

zbowling opened the pull request:
On platforms without fat binaries (anything but Darwin), we now build
and install stdlibs for each arch in it's own folder. Namely
lib/swift/\<platform\>/\<arch\> instead of lib/swift/\<platform\>.

(Split from #12955)

zbowling committed:
Fix multi-arch cross-compiling

On platforms without fat binaries (anything but Darwin), we now build
and install stdlibs for each arch in it's own folder. Namely
lib/swift/<platform>/<arch> instead of lib/swift/<platform>.

zbowling commented on cmake/modules/AddSwift.cmake:
```diff
@@ -327,7 +327,7 @@ function(_add_variant_link_flags)
 
   precondition(LFLAGS_SDK MESSAGE "Should specify an SDK")
-  precondition(LFLAGS_SDK MESSAGE "Should specify an architecture")
+  precondition(LFLAGS_ARCH MESSAGE "Should specify an architecture")
```
Fixed an unrelated copy/paste bug with this arg.

zbowling commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1068,8 +1065,16 @@ function(_add_swift_library_single target name)
-      "${SWIFT_NATIVE_SWIFT_TOOLS_PATH}/../lib/swift/${SWIFTLIB_SINGLE_SUBDIR}"
+      "${SWIFT_NATIVE_SWIFT_TOOLS_PATH}/../lib/swift/${SWIFTLIB_SINGLE_SUBDIR}")
+
+  if("${SWIFT_SDK_${SWIFTLIB_SINGLE_SDK}_OBJECT_FORMAT}" STREQUAL "ELF")
```
Not sure if I should gate on elf or gate on mach. @compnerd windows?

compnerd replied:
PE/COFF doesn't have fat binaries either.  I think that we should make this a Mach check.

gparker42 replied:
Line 1650 uses an `IS_DARWIN` check instead.

zbowling replied:
Yeah, I can switch to that.

jrose-apple commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1016,6 +1016,17 @@ static void getRuntimeLibraryPath(SmallVectorImpl<char> &runtimeLibPath,
 }
 
+/// On Darwin we link fat libs in the root of the library path and on other OS
+/// targets we look up libraries in the target arch's subfolder.
```
Nitpick: This comment doesn't really describe the function.

zbowling replied:
Ah true. What I was thinking when I changed where we used it and not what the function does.

jrose-apple reviewed (commented):
Driver parts look okay to me. I'll let Ross and others comment on the CMake.

gparker42 commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1631,47 +1647,108 @@ function(add_swift_library name)
-          # Note this thin library.
-          list(APPEND THIN_INPUT_TARGETS ${VARIANT_NAME})
+            if (SWIFTLIB_IS_STDLIB AND SWIFTLIB_STATIC)
+              # Add dependencies on the (not-yet-created) custom lipo target.
```
This comment can't be right. This is inside the non-Darwin no-lipo case.

zbowling replied:
Ah good catch. Copy-pasta.

compnerd commented:
I will try to take a closer look at this soon.  But, we should try to get this patch taken care of soon.  There is at least one patch from me (apple/swift#12480) and one from @troughton (apple/swift#13140) which is trying to address the problems caused by the fat vs thin binary handling in the build.

zbowling commented:
@compnerd Yeah, I see what you are doing in that other patch. I'll try to fix this one up really quick.

zbowling commented:
Fixed issues. can someone kick off the CI bot for me? I can't test locally this weekend.

phausler commented:
@swift-ci please test

phausler commented:
@swift-ci please test

phausler commented:
@swift-ci please test

zbowling commented:
not sure how to invoke @swift-ci so it compiles this PR with foundation using this PR together https://github.com/apple/swift-corelibs-foundation/pull/1354

gparker42 commented:
please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
doh. dependency in unit tests was broken. fixed

gparker42 commented:
please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
Funny linker error in foundation: 

```../buildbot_linux/foundation-linux-x86_64/Foundation/Foundation/NSIndexSet.swift.o: In function `_T010Foundation10NSIndexSetC21_enumerateWithOptionsSiSgAA013NSEnumerationF0V_AA8_NSRangeV5rangexm9paramTypeq_m06returnK0q_x_SpyAA8ObjCBoolVGtc5blocktr0_lFySicq_x_AOtccfU_ySicfU_AI_ytTg5':
../buildbot_linux/foundation-linux-x86_64/Foundation/Foundation/NSIndexSet.swift.o:(.text+0x2b2f): undefined reference to `_T0s17_AnyCollectionBoxC6_indexs01_a5IndexC0_psAD_p_s5Int64V8offsetBytF'```

Curious why this worked on a previous build. 

Demangle says it's: Swift._AnyCollectionBox._index(Swift._AnyIndexBox, offsetBy: Swift.Int64) -> Swift._AnyIndexBox

I'm wondering if this is supposed to get generated by libswiftSwiftOnoneSupport.so and it's not picking it up now.

zbowling commented:
updated this branch to current master. I wondering if this is just the branch foundation is on is out of sync with this one. should be ok to test again.

gparker42 commented:
please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
I think it's failing because the foundation's build script was still looking in the former lib directory too and it's picking up on an older version of the libswift(Core|Onone).so since the CI isn't doing a clean build. 

I pushed a fix for the foundation build script to be a bit more explicit about search directories and that should fix it. My cloud builder at google is back up I'm running a test there first.

zbowling commented:
Just passed on my build machine. Good to run tests here again.

zbowling commented:
Mac test passed. Linux should pass with the update in the PR I pushed.

zbowling commented:
@compnerd @troughton want to take a look now? 

@gparker42 @phausler thanks for kicking off CI builds for me tonight (hope to be able to do that myself eventually 😄 ). can I get one more test kick with that foundation pr again? Can probably just do Linux or both or whatever.

phausler commented:
please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
Looks like the linux CI builder is busted

gottesmm commented:
@zbowling This is when you raise the Mishal Signal... @shahmishal where are you, we need you! Raise the signal:

![](https://imgflip.com/i/20hqmn)

shahmishal commented:
@swift-ci test Linux

shahmishal commented:
We had a network issue, took the node offline and working on resolving issue soon.

zbowling commented:
@shahmishal this PR requires running the test with the PR in the foundation repo above :)

zbowling commented:
Looks like libswiftImageInspectionShared.a still gets stored in lib/swift_static/linux/ without an arch suffix still.

This issue relates to the issue that static stdlibs are only generated for host variant sdks instead of all sdks but done so in a way that actually broken (some libs are generated for other SDKs but not correctly). 

In generally things are pretty broken for static linking for several reasons on Linux (or really anything other than Darwin) but was hoping to fix all this all in another PR. 

It's a bit of hack how it does things in swift/stdlib/public/runtime/CMakeLists.txt today where the code:

1. doesn't generate swiftImageInspectionShared correctly when cross compiling wherein each additional target arch of the stdlib you compile causes the last version of that lib, regardless of target, to smash and replace the Linux one (like doing a Android or Fuchsia build on Linux with --build-swift-static-stdlib set). 
2. uses the static-executable-args.lnk file and which only gets copied into the Linux lib folder and not other target sdks
3. changes the way the shared version of swiftRuntime is generated oddly when you also make a static stdlib build as well making an additional shared library with just the ImageInspectionELF.cpp which seems like totally unnecessary overhead and makes an additional library that you sometimes have to deal with deploying. I'm not sure why we do that other than making it easier to to invoke add_swift_library once so it makes both static and shared versions instead of calling it twice with, once with STATIC and once with DYNAMIC with the two different file lists. ( @compnerd any idea? )

Part of me wishes we could kill the SWIFT_HOST_VARIANT_SDK variable usage out side of the root CMakeList.txt because it encourages bad behaviors where code can be created that doesn't deal with multiple SDKs being built that seems to work only when host=target. In theory we should very very rarely being doing anything specific for the host variant we are building on. Nearly everywhere in all the cmake files that variable is used, outside of the root CMakeList.txt file, I've found bugs with cross compiling that undermine or don't consider all the different parts of AddSwift.cmake's logic to handle building for each SDK.

phausler commented:
one consideration here for the cross-repo dependency; we could perhaps change the Foundation build script to look in both places (since it seems pretty reasonable to me) and that would free this commit from being paired to the sclf repo.

zbowling commented:
@phausler i did that earlier. unfortunately the CI bot didn't do a clean build and picked up an older version of the runtime libs and broke at link time. my last patch for the foundation side just made it explicit.

zbowling commented:
Looks like the haiku port broke a validation test with glibc.modulemap generation. Fixed it here (causing /// to added to paths in the sysroot on install tarballs) 

also fixed the resource directory linking test that failed (now knows about the arch directory).
 
the remaining test for static stdlib is still failing until we can fix the issues with static stdlibs I mentioned above. I can do something that fixes it in a way that just makes the test work but it static linking will still be fairly broken.

zbowling commented:
actually with a lot of @compnerd's elf changes, we don't even need this static stuff since static and dynamic elf image inspection is done the same way. fix incoming that just fixes everything. this will drop the additional .so we generate in static stdlib builds.

zbowling committed:
remove concept of swiftImageInspectionStatic/Shared libraries

zbowling commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -80,43 +80,34 @@ set(swift_runtime_library_compile_flags ${swift_runtime_compile_flags})
+    if(NOT "${sdk}" STREQUAL "LINUX" AND
+       NOT "${sdk}" STREQUAL "FREEBSD" AND
+       NOT "${sdk}" STREQUAL "ANDROID")
+      continue()
```
I believe only these SDKs will work with the static-executable-args.lnk file we have now. The arguments in that file are linux specific. We should in theory dynamically generate it like we do with the gen-static-stdlib script instead.

phausler commented:
please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
Pushed a fix for the one test that was failing. The other one I don't understand why it's failing. I merged master in the hope that it fixes it in the next test.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -486,11 +486,8 @@ function(_add_swift_lipo_target)
-        OUTPUT "${LIPO_OUTPUT}"
-        DEPENDS ${source_targets})
+    message(FATAL_ERROR
+        "It's not possible to build a universal binary for a non-Darwin platform")
```
This is a MachO feature, not a Darwin feature I believe.

zbowling replied:
Darwin being the only platform that has a lipo like tool though to actually make multi-arch binaries :) 

Technically there is the FatELF thing but it's niche and pretty much dead.

compnerd replied:
There also have been other projects to run MachO binaries on Linux.  In such a case, a `lipo` tool could work on Linux just as well.  It really is something which is part of the MachO binary format, and it just happens that Darwin has the most stable support for MachO binaries.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1055,8 +1052,17 @@ function(_add_swift_library_single target name)
+    list(APPEND library_search_directories
       "${SWIFT_NATIVE_SWIFT_TOOLS_PATH}/../lib/swift/${SWIFT_SDK_${SWIFTLIB_SINGLE_SDK}_LIB_SUBDIR}")
+  else()
+    # No fat binaries on other archs so search in an arch specific directory
```
s/archs/targets/ ?

zbowling replied:
good catch. English is my first language but I'm still terrible at it. Probably why I stuck to math in school.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1631,47 +1649,107 @@ function(add_swift_library name)
+              endforeach()
             endif()
-          endforeach()
+            # Note this thin library.
```
"this is a thin library"?

zbowling replied:
I'm not sure. Was there before so I didn't break it. It's darwin specific I believe so don't recall what it's for.

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1631,47 +1649,107 @@ function(add_swift_library name)
-                add_dependencies("${VARIANT_NAME}-static"
-                  "${DEP}-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-static")
+                add_dependencies(${VARIANT_NAME}
+                  "${DEP}-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
```
Doesn't this convert from a static link to a dynamic link?

zbowling replied:
The static case is handled in the if statement below (order was reversed and diff is confused)

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -1631,47 +1649,107 @@ function(add_swift_library name)
+                      WORLD_READ WORLD_EXECUTE)
+                  set(install_libpath
+                    "${SWIFTLIB_DIR}/${SWIFT_SDK_${sdk}_LIB_SUBDIR}/${arch}/${CMAKE_SHARED_LIBRARY_PREFIX}${name}${CMAKE_SHARED_LIBRARY_SUFFIX}")
+
```
Unnecessary whitespace

zbowling replied:
Was preserved from before. I just moved the code around. Doesn't look right in github but it looks wrapped properly in the code.

compnerd commented on lib/Driver/ToolChains.cpp:
```diff
@@ -1016,6 +1016,17 @@ static void getRuntimeLibraryPath(SmallVectorImpl<char> &runtimeLibPath,
 
+/// Get the runtime library link path with the target triple's specific arch
+/// library folder.
+static void getRuntimeLibraryPathWithArch(SmallVectorImpl<char> &runtimeLibPath,
```
Can you call this `getRuntimeSharedLibraryPathWithArch` to match `getRuntimeStaticLibraryPathWithArch`?

compnerd commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -45,10 +45,19 @@ static void updateRuntimeLibraryPath(SearchPathOptions &SearchPathOpts,
+    SearchPathOpts.RuntimeLibraryPath = LibPath.str();
+  }
 }
 
```
Why not sink the LibPath calculation into the `Triple.isOSDarwin()` or `else` case respectively?

zbowling replied:
Sure. I'll look at writing this be more clear with one branch instead of two. I was over optimizing trying to avoid creating two string allocs for the non-Darwin path at the sake of some clarity. 

Basically:

```
if Darwin:
  RuntimeLibraryPath = /lib/swift/(platform)
else:
  RuntimeLibraryPath = /lib/swift/(platform)/(arch)

RuntimeLibraryImportPath = /lib/swift/(platform)/(arch)
```

compnerd commented:
Please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
I think one of the tests failed because it's not cleaning. A previous version created a symlink in an unfortunate spot that will confuse the "RUN: ln -s" statements in the test. 

The other test I'm not sure why it fails. It's not failing that way locally on my Ubuntu 14.04 machine.

zbowling commented on validation-test/execution/interpret-with-dependencies-linux.swift:
```diff
@@ -8,9 +8,9 @@
+// RUN: %empty-directory(%t/rsrc/%target-sdk-name/%target-cpu)
 // RUN: ln -s %t/libabc.so %t/rsrc/%target-sdk-name/%target-cpu
-// RUN: ln -s %platform-module-dir/../* %t/rsrc/%target-sdk-name/%target-cpu
+// RUN: ln -s %platform-module-dir/../%target_cpu/* %t/rsrc/%target-sdk-name/%target-cpu
```
doh. %target-cpu and not %target_cpu

zbowling commented:
Seems like a regression to something discussed in https://github.com/apple/swift/pull/282

zbowling commented:
squashed and rebased this branch so I can update my fuchsia branch based on these commits

Rostepher commented:
@swift-ci please smoke test

zbowling commented:
Please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

compnerd commented:
Please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please test

zbowling commented:
Please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please smoke test linux

zbowling commented:
I'm not sure what this test is doing that is failing on the linux bot. The module maps generated are working on debain machines I've tested just fine but the test fails.

furkanozbay commented:
@swift-ci please test

CodaFi commented:
Drive-by log restoration. Don’t mind me

CodaFi commented:
Please test with the following pull request:
apple/swift-corelibs-foundation#1354

@swift-ci please smoke test Linux platform

compnerd commented:
Do we want to split this up and get in part of the patch earlier?  The build portion would be extremely useful for me as it allows me to build just part of the windows targets (e.g. just the x86_64 or ARM target).

zbowling commented:
@compnerd it's very intertwined and already split up a lot so far. going to try and investigate this last test that is failing. it's some kind of intermediate thing testing some kind of module metadata that I don't really understand or something but the final product seems to working fine on all archs so it's weird.

alblue commented:
@swift-ci please test linux

zbowling commented:
I have some changes to swiftpm, libdispatch, and xctest (in additional to the current foundation change already) that go along with this change to teach them about the linux arch folder.  Still not sure what is broken this unit test though.

lanza commented:
@zbowling Hey Zac, is this still being worked on? We could use some of these changes.

CodaFi commented:
Closing this due to age and inactivity, both here and on the corresponding Foundation PR.
